chia/wallet/nft_wallet/uncurry_nft.py

```diff
@@ -0,0 +1,98 @@
+from chia.types.blockchain_format.program import Program
+from chia.wallet.puzzles.load_clvm import load_clvm
+
+SINGLETON_TOP_LAYER_MOD = load_clvm("singleton_top_layer_v1_1.clvm")
+NFT_MOD = load_clvm("nft_innerpuz.clvm")
+
+
+class UncurriedNFT:
+    """
+    A simple solution for uncurry NFT puzzle.
+    Initial the class with a full NFT puzzle, it will do a deep uncurry.
+    This is the only place you need to change after modified the Chialisp curried parameters.
+    """
+
+    matched: bool
+    """If the puzzle is a NFT puzzle"""
+
+    nft_mod_hash: Program
+    """NFT module hash"""
+
+    singleton_struct: Program
+    """
+    Singleton struct
+    [singleton_mod_hash, singleton_launcher_id, launcher_puzhash]
+    """
+    singleton_mod_hash: Program
+    singleton_launcher_id: Program
+    launcher_puzhash: Program
+
+    owner_did: Program
+    """Owner's DID"""
+
+    transfer_program_hash: Program
+    """Puzzle hash of the transfer program"""
+
+    transfer_program_curry_params: Program
+    """
+    Curried parameters of the transfer program
+    [royalty_address, trade_price_percentage, settlement_mod_hash, cat_mod_hash]
+    """
+    royalty_address: Program
+    trade_price_percentage: Program
+    settlement_mod_hash: Program
+    cat_mod_hash: Program
+
+    metadata: Program
+    """
+    NFT metadata
+    [("u", data_uris), ("h", data_hash)]
+    """
+    data_uris: Program
+    data_hash: Program
+
+    def __init__(self, puzzle: Program, raise_exception: bool = False):
```

Comment:
There has not been a team consensus on this but I strongly encourage that every class be a `@dataclass` and not use `.__post_init__()` to add back complexity to the constructor.  Rather, have a helper function or `@classmethod` that does the fancy calculations needed to figure out the attributes of the to-be-created class.